### PR TITLE
[SPARK-11496][GRAPHX] Parallel implementation of personalized pagerank

### PR DESCRIPTION
**Pull request by moustaki**
(Updated version of [PR-9457](https://github.com/apache/spark/pull/9457), rebased on latest Spark master, and using mllib-local).

This implements a parallel version of personalized pagerank, which runs all propagations for a list of source vertices in parallel.

I ran a few benchmarks on the full [DBpedia](http://dbpedia.org/) graph. When running personalized pagerank for only one source node, the existing implementation is twice as fast as the parallel one (because of the SparseVector overhead). However for 10 source nodes, the parallel implementation is four times as fast. When increasing the number of source nodes, this difference becomes even greater.  

image

### CODE REVIEW

File: graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala [hunk 1]
```diff
@@ -19,8 +19,11 @@ package org.apache.spark.graphx.lib
 
 import scala.reflect.ClassTag
 
-import org.apache.spark.graphx._
+import breeze.linalg.{Vector => BV}
+
+import org.apache.spark.ml.linalg.{Vector, Vectors}
```
Comment (dbtsai) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala]: move `org.apache.spark.ml.linalg.{Vector, Vectors}` down, and move `org.apache.spark.graphx._` up. We have an automated system detecting if the import is alphabetical, and that is failing the build.

**Comment by dbtsai** [comment]
Jenkins, retest this please

File: graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala [hunk 2]
```diff
@@ -163,6 +166,85 @@ object PageRank extends Logging {
...
+          } else {
+            zero
+          }
+          popActivations :+ resetActivations
+        }.cache()
+
+      rankGraph.edges.foreachPartition(x => {}) // also materializes rankGraph.vertices
+      prevRankGraph.vertices.unpersist(false)
+      prevRankGraph.edges.unpersist(false)
+
```
Comment (dbtsai) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala]: add `sourcesInitMapBC.destory(false)`

File: graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala [hunk 3]
```diff
@@ -163,6 +166,85 @@ object PageRank extends Logging {
...
+      val prevRankGraph = rankGraph
+      // Propagates the message along outbound edges
+      // and adding start nodes back in with activation resetProb
+      val rankUpdates = rankGraph.aggregateMessages[BV[Double]](
+        ctx => ctx.sendToDst(ctx.srcAttr :* ctx.attr),
+        (a : BV[Double], b : BV[Double]) => a :+ b, TripletFields.Src)
+
+      rankGraph = rankGraph.joinVertices(rankUpdates) {
+        (vid, oldRank, msgSum) =>
+          val popActivations: BV[Double] = msgSum :* (1.0 - resetProb)
```
Comment (dbtsai) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala]: We found that `breeze` is fairly slow by doing this operation. Is it possible to use native spark vector, and use our linear algebra package?

File: graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala [hunk 4]
```diff
@@ -163,6 +166,85 @@ object PageRank extends Logging {
...
+      prevRankGraph.vertices.unpersist(false)
+      prevRankGraph.edges.unpersist(false)
+
+      logInfo(s"Parallel Personalized PageRank finished iteration $i.")
+
+      i += 1
+    }
+
+    rankGraph
+      .mapVertices { (vid, attr) =>
```
Comment (dbtsai) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala]: move it up one line.

**Comment by dbtsai** [comment]
Some minor issues, and LGTM. Thanks.

File: graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala (hunk 3 above)
Comment (moustaki) [inline graphx/src/main/scala/org/apache/spark/graphx/lib/PageRank.scala]: Chatted offline, but decided to not switch to native Spark operations, as it doesn't support adding sparse vectors, which the above does

**Comment by dbtsai** [comment]
Jenkins, retest this please

**Comment by dbtsai** [comment]
Merged into master. Great work! Thanks.

**Pull request closed without merging**